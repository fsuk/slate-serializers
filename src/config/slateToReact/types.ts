import { ReactNode } from 'react'

interface MarkTagTransform {
  [key: string]: ({
    node,
    attribs,
  }: {
    node?: any
    attribs?: { [key: string]: string }
  }) => ReactNode
}

interface ElementTagTransform {
  [key: string]: ({
    node,
    attribs,
    children,
  }: {
    node?: any
    attribs?: { [key: string]: string }
    children?: ReactNode
  }) => ReactNode
}

export interface Config {
  markMap: { [key: string]: string[] }
  elementMap: { [key: string]: string }
  elementStyleMap?: {
    [key: string]: string
  }
  markStyleMap?: {
    [key: string]: string
  }
  markTransforms?: MarkTagTransform
  elementTransforms: ElementTagTransform
  defaultTag?: string
  encodeEntities?: boolean
  alwaysEncodeBreakingEntities?: boolean
  alwaysEncodeCodeEntities?: boolean
  convertLineBreakToBr?: boolean
}
