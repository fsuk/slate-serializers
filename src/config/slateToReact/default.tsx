import React, { ReactNode } from "react"
import { Config } from './types'
import { nanoid } from 'nanoid'

const BlockQuote = ({ children }: { children: ReactNode }) =>
  <blockquote>
    <p>
      {children}
    </p>
  </blockquote>

export const config: Config = {
  elementTransforms: {
    quote: ({ children }) => {
      return <BlockQuote key={nanoid()}>{children}</BlockQuote>
    },
    link: ({ node, children = [] }) => {
      const attrs: any = {}
      if (node.newTab) {
        attrs.target = '_blank'
      }
      return <a
        key={nanoid()}
        href={node.url}
        {...attrs}
      >{children}</a>
    },
  },
}
