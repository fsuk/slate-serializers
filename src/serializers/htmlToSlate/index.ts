import { jsx } from 'slate-hyperscript'
import { Parser, ElementType } from 'htmlparser2'
import { ChildNode, DomHandler, Element, Text } from 'domhandler'
import { getAttributeValue, getChildren, getName, textContent } from 'domutils'
import { removeLineBreaks } from '../../utilities'

interface ItagMap {
  [key: string]: (a?: Element) => object
}

const ELEMENT_TAGS: ItagMap = {
  a: (el) => ({
    type: 'link',
    newTab: el && getAttributeValue(el, 'target') === '_blank',
    url: el && getAttributeValue(el, 'href'),
  }),
  blockquote: () => ({ type: 'blockquote' }),
  h1: () => ({ type: 'h1' }),
  h2: () => ({ type: 'h2' }),
  h3: () => ({ type: 'h3' }),
  h4: () => ({ type: 'h4' }),
  h5: () => ({ type: 'h5' }),
  h6: () => ({ type: 'h6' }),
  li: () => ({ type: 'li' }),
  ol: () => ({ type: 'ol' }),
  p: () => ({ type: 'p' }),
  pre: () => ({ type: 'code' }),
  ul: () => ({ type: 'ul' }),
}

const TEXT_TAGS: ItagMap = {
  code: () => ({ code: true }),
  del: () => ({ strikethrough: true }),
  em: () => ({ italic: true }),
  i: () => ({ italic: true }),
  s: () => ({ strikethrough: true }),
  strong: () => ({ bold: true }),
  u: () => ({ underline: true }),
}

const deserialize = (el: ChildNode, index?: number): any => {
  if (el.type !== ElementType.Tag && el.type !== ElementType.Text) {
    return null
  }
  let parent = el as Element
  if (getName(parent) === 'br') {
    return '\n'
  }

  const nodeName = getName(parent)
  const nodeFirstChildName = parent.childNodes && parent.childNodes[0] && getName(parent.childNodes[0] as Element)
  const nodeFirstParentName = parent.parent && getName(parent.parent as Element)

  let children = parent.childNodes ? Array.from(parent.childNodes).map(deserialize).flat() : []

  if (getName(parent) === 'body') {
    return jsx('fragment', {}, children)
  }

  if (
    (nodeName === 'pre' && nodeFirstChildName === 'code') ||
    (nodeName === 'code' && nodeFirstChildName === 'pre')
  ) {
    const attrs = TEXT_TAGS['code'](parent)
    return [jsx('text', attrs, textContent(getChildren(parent.childNodes[0])))]
  }

  if (ELEMENT_TAGS[nodeName]) {
    const attrs = ELEMENT_TAGS[nodeName](parent)
    return jsx('element', attrs, children)
  }

  if (el.type === ElementType.Text) {
    if (textContent(el).trim() === '') {
      return null
    }
    const name = nodeName || nodeFirstParentName || ''
    const attrs = TEXT_TAGS[name] ? TEXT_TAGS[name](parent) : {}
    return [jsx('text', {...attrs, text: textContent(el)}, children)]
  }

  return children
}

export const htmlToSlate = (html: string) => {
  let slateContent
  const handler = new DomHandler((error, dom) => {
    if (error) {
      // Handle error
    } else {
      // Parsing completed, do something
      slateContent = dom
        .map((node) => deserialize(node)) // run the deserializer
        .filter((element) => element) // filter out null elements
        .map((element) => { // ensure all top level elements have a children property
          if (!element.children) {
            return {
              children: element
            }
          }
          return element
        })
    }
  })
  const parser = new Parser(handler)
  parser.write(removeLineBreaks(html))
  parser.end()
  return slateContent
}
