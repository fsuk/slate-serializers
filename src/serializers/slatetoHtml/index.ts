import { escape } from 'html-escaper'
import { Text } from 'slate'

export const slateToHtml = (node: any[]) => {
  const slateNode = { children: node }
  return slateNodeToHtml(slateNode)
}

const slateNodeToHtml = (node: any) => {
  if (Text.isText(node)) {
    let str = escape(node.text)
    if ((node as any).bold) {
      str = `<strong>${str}</strong>`
    }
    return str
  }

  const children = node.children && node.children.map((n: any[]) => slateNodeToHtml(n)).join('')

  switch (node.type) {
    case 'p':
      return `<p>${children}</p>`
    case 'h1':
      return `<h1>${children}</h1>`
    case 'h2':
      return `<h2>${children}</h2>`
    case 'quote':
      return `<blockquote><p>${children}</p></blockquote>`
    case 'paragraph':
      return `<p>${children}</p>`
    case 'ul':
      return `<ul>${children}</ul>`
    case 'li':
      return `<li>${children}</li>`
    case 'link':
      return `<a href="${escape(node.url)}">${children}</a>`
    default:
      return children
  }
}
