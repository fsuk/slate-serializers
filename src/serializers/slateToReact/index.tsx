import React, { ReactElement,  JSXElementConstructor } from 'react'
import { Element, isTag, Text } from 'domhandler'
import { getName, textContent } from 'domutils'
import { nanoid } from 'nanoid'

import { config as defaultConfig } from '../../config/slateToDom/default'
import { Config } from '../../config/slateToDom/types'
import { slateToReactConfig, type SlateToReactConfig } from '../../react'
import { convertSlate } from '../../utilities/convert-slate'

interface ISlateToReact {
  node: any[]
  config?: Config
  reactConfig?: SlateToReactConfig
}

export const SlateToReact = ({
  node,
  config = defaultConfig,
  reactConfig = slateToReactConfig
}: ISlateToReact) => {
  if (!Array.isArray(node)) {
    return <></>
  }
  const document = node.map((n, index) => convertSlate({
    node: n,
    ...{
      config,
      elementTransforms: {},
    },
    isLastNodeInDocument: index === node.length - 1,
    customElementTransforms: reactConfig.elementTransforms,
    transformText: (text) => transformText(text),
    transformElement: (element) => {
      return domElementToReactElement(element)
    },
    wrapChildren: (children) => children
  }))
  return document as any
}

const transformText = (text: Text | Element) => {
  if (isTag(text as Element)) {
    return React.createElement(getName(text as Element), {}, textContent(text as Element))
  }
  return <>{textContent(text as Text)}</>
}

const domElementToReactElement = (element: Element): ReactElement<any, string | JSXElementConstructor<any>> => {
  return React.createElement(
    getName(element),
    {
      /* Ensure all React elements have a unique key - is there a better way to do this? */
      key: nanoid(),
      /* Provide all attributes as props */
      ...element.attribs,
      /* Convert key names for JSX compatibility */
      ...(element.attribs?.class && { className: element.attribs?.class }),
    },
    element.children as any
  )
}
