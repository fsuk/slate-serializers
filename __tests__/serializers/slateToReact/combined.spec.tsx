import React from 'react'
import renderer from 'react-test-renderer';
import { fixtures as elementFixtures } from '../combined/fixtures/elementTags'
import { fixtures as textFixtures } from '../combined/fixtures/textTags'
import { SlateToReact } from '../../../src/serializers/slateToReact'

describe('Slate JSON to React transforms', () => {
  describe('Element tags', () => {
    const fixtures = elementFixtures
    for (const fixture of fixtures) {
      it(`${fixture.name}`, () => {
        // console.log((<SlateToReact node={fixture.slate} />).props, (<SlateToReact node={fixture.slate} />).props.node.map((node: any) => node.children))
        const tree = renderer
          .create(<SlateToReact node={fixture.slate} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      })
    }
  })
  describe('Text tags', () => {
    const fixtures = textFixtures
    for (const fixture of fixtures) {
      it(`${fixture.name}`, () => {
        const tree = renderer
          .create(<SlateToReact node={fixture.slate} />)
          .toJSON();
        expect(tree).toMatchSnapshot();
      })
    }
  })
})