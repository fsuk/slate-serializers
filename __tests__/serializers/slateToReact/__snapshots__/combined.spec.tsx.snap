// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slate JSON to React transforms Element tags blockquote 1`] = `
<blockquote>
  Blockquote
</blockquote>
`;

exports[`Slate JSON to React transforms Element tags headings 1`] = `
[
  <h1>
    Heading 1
  </h1>,
  <h2>
    Heading 2
  </h2>,
  <h3>
    Heading 3
  </h3>,
  <h4>
    Heading 4
  </h4>,
  <h5>
    Heading 5
  </h5>,
  <h6>
    Heading 6
  </h6>,
]
`;

exports[`Slate JSON to React transforms Element tags link 1`] = `
<a>
  Slate Serializers | GitHub
</a>
`;

exports[`Slate JSON to React transforms Element tags nested unordered list 1`] = `
<ul>
  <li>
    Item 1
    <ul>
      <li>
        Nested item 1
      </li>
      <li>
        Nested item 2
      </li>
    </ul>
  </li>
  <li>
    Item 2
  </li>
</ul>
`;

exports[`Slate JSON to React transforms Element tags ordered list 1`] = `
<ol>
  <li>
    Item 1
  </li>
  <li>
    Item 2
  </li>
</ol>
`;

exports[`Slate JSON to React transforms Element tags paragraph 1`] = `
<p>
  Paragraph 1
</p>
`;

exports[`Slate JSON to React transforms Element tags unordered list 1`] = `
<ul>
  <li>
    Item 1
  </li>
  <li>
    Item 2
  </li>
</ul>
`;

exports[`Slate JSON to React transforms Text tags idiomatic text 1`] = `
<i>
  Idiomatic Text
</i>
`;

exports[`Slate JSON to React transforms Text tags pre 1`] = `
<pre>
  Pre
</pre>
`;

exports[`Slate JSON to React transforms Text tags strikethrough 1`] = `
<s>
  Strikethrough
</s>
`;

exports[`Slate JSON to React transforms Text tags strong 1`] = `
<strong>
  Bold
</strong>
`;

exports[`Slate JSON to React transforms Text tags strong and idiomatic text 1`] = `
<strong>
  Strong and Idiomatic Text
</strong>
`;

exports[`Slate JSON to React transforms Text tags strong and unarticulated annotation 1`] = `
<strong>
  Strong and Unarticulated Annotation
</strong>
`;

exports[`Slate JSON to React transforms Text tags strong in paragraph 1`] = `
<p>
  <strong>
    Bold
  </strong>
</p>
`;

exports[`Slate JSON to React transforms Text tags strong mid-sentence 1`] = `
[
  "Some ",
  <strong>
    bold text
  </strong>,
  " in a sentence.",
]
`;

exports[`Slate JSON to React transforms Text tags unarticulated annotation 1`] = `
<u>
  Unarticulated Annotation
</u>
`;
